File: cpp/PseudoClasses.cpp
#include "PseudoClasses.hpp"

namespace margelo::nitro::cssnitro {

    // Initialize the static map
    std::unordered_map<std::string, PseudoClassState> PseudoClasses::states;

    bool PseudoClasses::get(const std::string &key, PseudoClassType type,
                            reactnativecss::Effect::GetProxy &get) {
        // Find or create the state for this key
        auto stateIt = states.find(key);
        if (stateIt == states.end()) {
            // Key doesn't exist, create it
            PseudoClassState newState;
            states[key] = newState;
            stateIt = states.find(key);
        }

        auto &state = stateIt->second;

        // Get the appropriate observable based on type
        std::optional<std::shared_ptr<reactnativecss::Observable<bool>>> *observablePtr = nullptr;

        // Select the correct observable based on the PseudoClassType
        switch (type) {
            case PseudoClassType::ACTIVE:
                observablePtr = &state.active;
                break;
            case PseudoClassType::HOVER:
                observablePtr = &state.hover;
                break;
            case PseudoClassType::FOCUS:
                observablePtr = &state.focus;
                break;
        }

        // Safety check - should never happen but guards against warnings
        if (observablePtr == nullptr) {
            return false;
        }

        // If the observable doesn't exist, create it with default value of false
        if (!observablePtr->has_value()) {
            *observablePtr = reactnativecss::Observable<bool>::create(false);
        }

        // Subscribe to the observable and return its value
        return get(*observablePtr->value());
    }

    void PseudoClasses::set(const std::string &key, PseudoClassType type, bool value) {
        // Find or create the state for this key
        auto stateIt = states.find(key);
        if (stateIt == states.end()) {
            // Key doesn't exist, create it
            PseudoClassState newState;
            states[key] = newState;
            stateIt = states.find(key);
        }

        auto &state = stateIt->second;

        // Get the appropriate observable based on type
        std::optional<std::shared_ptr<reactnativecss::Observable<bool>>> *observablePtr = nullptr;

        switch (type) {
            case PseudoClassType::ACTIVE:
                observablePtr = &state.active;
                break;
            case PseudoClassType::HOVER:
                observablePtr = &state.hover;
                break;
            case PseudoClassType::FOCUS:
                observablePtr = &state.focus;
                break;
        }

        // Safety check - should never happen but guards against warnings
        if (observablePtr == nullptr) {
            return;
        }

        // If the observable doesn't exist, create it
        if (!observablePtr->has_value()) {
            *observablePtr = reactnativecss::Observable<bool>::create(value);
        } else {
            // Update the existing observable
            observablePtr->value()->set(value);
        }
    }

    void PseudoClasses::remove(const std::string &key) {
        // Remove the key from the map
        states.erase(key);
    }

} // namespace margelo::nitro::cssnitro

File: src/native-ts/PseudoClasses.ts
/**
 * PseudoClasses: Pseudo-class state management
 * Converted from cpp/PseudoClasses.cpp
 */

import type { PseudoClassType } from "../specs/StyleRegistry";
import type { GetProxy } from "./Effect";
import { Observable } from "./Observable";

interface PseudoClassState {
  active?: Observable<boolean>;
  hover?: Observable<boolean>;
  focus?: Observable<boolean>;
}

const states = new Map<string, PseudoClassState>();

function ensureState(key: string): PseudoClassState {
  let state = states.get(key);
  if (!state) {
    state = {};
    states.set(key, state);
  }
  return state;
}

export const PseudoClasses = {
  get(key: string, type: PseudoClassType, get: GetProxy): boolean {
    // Find or create the state for this key
    const state = ensureState(key);

    // Get the appropriate observable based on type
    let observableRef: Observable<boolean> | undefined;

    switch (type) {
      case "active":
        observableRef = state.active;
        break;
      case "hover":
        observableRef = state.hover;
        break;
      case "focus":
        observableRef = state.focus;
        break;
    }

    // If the observable doesn't exist, create it with default value of false
    if (!observableRef) {
      observableRef = Observable.create(false);
      switch (type) {
        case "active":
          state.active = observableRef;
          break;
        case "hover":
          state.hover = observableRef;
          break;
        case "focus":
          state.focus = observableRef;
          break;
      }
    }

    // Subscribe to the observable and return its value
    return get(observableRef);
  },

  set(key: string, type: PseudoClassType, value: boolean): void {
    // Find or create the state for this key
    const state = ensureState(key);

    // Get the appropriate observable based on type
    let observableRef: Observable<boolean> | undefined;

    switch (type) {
      case "active":
        observableRef = state.active;
        break;
      case "hover":
        observableRef = state.hover;
        break;
      case "focus":
        observableRef = state.focus;
        break;
    }

    // If the observable doesn't exist, create it
    if (!observableRef) {
      observableRef = Observable.create(value);
      switch (type) {
        case "active":
          state.active = observableRef;
          break;
        case "hover":
          state.hover = observableRef;
          break;
        case "focus":
          state.focus = observableRef;
          break;
      }
    } else {
      // Update the existing observable
      observableRef.set(value);
    }
  },

  remove(key: string): void {
    states.delete(key);
  },
};
